### ISSUE TITLE
[DRC-511] [Bug] recce server client-side exception has occurred

### ISSUE THREAD

**Issue by tomoki-takahashi-oisix** [DataRecce/recce#344]
### Current Behavior

I ran the `recce server` command, but I encountered the following error:

#### Chrome Browser
```
Application error: a client-side exception has occurred (see the browser console for more information).
```

#### Chrome DevTools
```
TypeError: Cannot read properties of undefined (reading 'children')
    at page-07217532d4bd1739.js:1:6996
    at page-07217532d4bd1739.js:1:8085
    at Object.r5 [as useMemo] (ce06bfec-6dfc7c0b1542aeb5.js:1:46765)
    at t.useMemo (584-f6e32ff02d123e11.js:1:111475)
    at ew (page-07217532d4bd1739.js:1:6125)
    at rk (ce06bfec-6dfc7c0b1542aeb5.js:1:40398)
    at lI (ce06bfec-6dfc7c0b1542aeb5.js:1:59205)
    at iB (ce06bfec-6dfc7c0b1542aeb5.js:1:117304)
    at o4 (ce06bfec-6dfc7c0b1542aeb5.js:1:94660)
    at ce06bfec-6dfc7c0b1542aeb5.js:1:94482
window.console.error @ 584-f6e32ff02d123e11.js:1
overrideMethod @ console.js:273
lE @ ce06bfec-6dfc7c0b1542aeb5.js:1
n.callback @ ce06bfec-6dfc7c0b1542aeb5.js:1
ng @ ce06bfec-6dfc7c0b1542aeb5.js:1
ny @ ce06bfec-6dfc7c0b1542aeb5.js:1
a$ @ ce06bfec-6dfc7c0b1542aeb5.js:1
aV @ ce06bfec-6dfc7c0b1542aeb5.js:1
a2 @ ce06bfec-6dfc7c0b1542aeb5.js:1
aV @ ce06bfec-6dfc7c0b1542aeb5.js:1
a2 @ ce06bfec-6dfc7c0b1542aeb5.js:1
aV @ ce06bfec-6dfc7c0b1542aeb5.js:1
a2 @ ce06bfec-6dfc7c0b1542aeb5.js:1
aV @ ce06bfec-6dfc7c0b1542aeb5.js:1
a2 @ ce06bfec-6dfc7c0b1542aeb5.js:1
aV @ ce06bfec-6dfc7c0b1542aeb5.js:1
a2 @ ce06bfec-6dfc7c0b1542aeb5.js:1
aV @ ce06bfec-6dfc7c0b1542aeb5.js:1
a2 @ ce06bfec-6dfc7c0b1542aeb5.js:1
aV @ ce06bfec-6dfc7c0b1542aeb5.js:1
a2 @ ce06bfec-6dfc7c0b1542aeb5.js:1
aV @ ce06bfec-6dfc7c0b1542aeb5.js:1
a2 @ ce06bfec-6dfc7c0b1542aeb5.js:1
aV @ ce06bfec-6dfc7c0b1542aeb5.js:1
a2 @ ce06bfec-6dfc7c0b1542aeb5.js:1
aV @ ce06bfec-6dfc7c0b1542aeb5.js:1
（匿名） @ ce06bfec-6dfc7c0b1542aeb5.js:1
o7 @ ce06bfec-6dfc7c0b1542aeb5.js:1
n2 @ ce06bfec-6dfc7c0b1542aeb5.js:1
n4 @ ce06bfec-6dfc7c0b1542aeb5.js:1
（匿名） @ ce06bfec-6dfc7c0b1542aeb5.js:1
setTimeout（非同期）
（匿名） @ 579-936b34f5c3294de9.js:37
y @ 579-936b34f5c3294de9.js:37
b @ 579-936b34f5c3294de9.js:37
a @ 579-936b34f5c3294de9.js:37
batch @ 579-936b34f5c3294de9.js:37
F @ 579-936b34f5c3294de9.js:37
setData @ 579-936b34f5c3294de9.js:37
onSuccess @ 579-936b34f5c3294de9.js:37
m @ 579-936b34f5c3294de9.js:37
Promise.then（非同期）
b @ 579-936b34f5c3294de9.js:37
c @ 579-936b34f5c3294de9.js:37
fetch @ 579-936b34f5c3294de9.js:37
U @ 579-936b34f5c3294de9.js:37
onSubscribe @ 579-936b34f5c3294de9.js:37
subscribe @ 579-936b34f5c3294de9.js:37
（匿名） @ 579-936b34f5c3294de9.js:37
rI @ ce06bfec-6dfc7c0b1542aeb5.js:1
aI @ ce06bfec-6dfc7c0b1542aeb5.js:1
a3 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a8 @ ce06bfec-6dfc7c0b1542aeb5.js:1
a5 @ ce06bfec-6dfc7c0b1542aeb5.js:1
ie @ ce06bfec-6dfc7c0b1542aeb5.js:1
n2 @ ce06bfec-6dfc7c0b1542aeb5.js:1
（匿名） @ ce06bfec-6dfc7c0b1542aeb5.js:1
o7 @ ce06bfec-6dfc7c0b1542aeb5.js:1
oH @ ce06bfec-6dfc7c0b1542aeb5.js:1
oj @ ce06bfec-6dfc7c0b1542aeb5.js:1
M @ 584-f6e32ff02d123e11.js:1
他 1 件のフレームを表示
一部表示
```

### Expected Behavior

Ensure that the server can be started using the 'recce server' command and that the application can be accessed without any errors.

### Steps To Reproduce

```
$ dbt docs generate --target-path target-base/  -t prod 
$ dbt docs generate
$ recce server
(Since there were errors in the Chrome browser, I opened the DevTools to investigate.)
```

### Environment

- recce: 0.22.1
- OS: MacOS 14.4.1
- Python: 3.11.9
- Data Warehouse: snowflake
- dbt: 1.7.1

<sub>[DRC-511](https://linear.app/piperider/issue/DRC-511/[bug]-recce-server-client-side-exception-has-occurred)</sub>

**Comment by popcornylu** [comment]
Hi @tomoki-takahashi-oisix, thanks for your bug report.

Could you please try the tutorial https://datarecce.io/docs/get-started-jaffle-shop/ , see if you can reproduce this error. I can successfully launch the recce server without error in the latest version (v0.22.2)
![image](https://github.com/DataRecce/recce/assets/860633/4fdf0525-27b1-4308-bc53-e9f05dbd32b8)

It would be helpful for us to investigate this issue if
1. Provide a reproducible way based on the 5-min tutorial 
2. or send the `target` and `target-base` in your project to dev+datarecce@infuseai.io if there is no privacy concern.

**Comment by kentwelcome** [comment]
### Finding

1. Error message `TypeError: Cannot read properties of undefined (reading 'children')` means we try to access `xxx.children`; however, the `xxx` was undefined.
2. The error was triggered at file `page-07217532d4bd1739.js:1:6996`. The original code will be at [lineage.ts:141](https://github.com/DataRecce/recce/blob/dc896d4b24f4d341334e2858cae2af2c60d52b2c/js/src/components/lineage/lineage.ts#L141)

   ![CleanShot 2024-06-19 at 10.42.56.png](https://uploads.linear.app/934ced25-db7b-45c5-ba31-730290f15f96/a446d24d-1711-4e14-ab57-5bc77222acd2/54cafd44-1af7-487d-ae41-3e95a4c99661?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzkzNGNlZDI1LWRiN2ItNDVjNS1iYTMxLTczMDI5MGYxNWY5Ni9hNDQ2ZDI0ZC0xNzExLTRlMTQtYWI1Ny01YmM3NzIyMmFjZDIvNTRjYWZkNDQtMWFmNy00ODdkLWFlNDEtM2U5NWE0Yzk5NjYxIiwiaWF0IjoxNzE4NzY3MzUyLCJleHAiOjE3NTAzMDMzNTJ9.q7WRXHESj6L2GbEC3QE5dvkN5SxRer9h4dhRux2IOwI)


3. This issue could happen in the following conditions
   1. The `parentNode` was undefined 
   2. According to `const parentNode = nodes[parent];`, the `nodes[parent]` should be undefined 
   3. Due to `nodes[key] = buildNode(key, "base");`, the value of `parent` should not be included in `base.parent_map` 

```
for (const [key, parents] of Object.entries(base.parent_map)) {
    nodes[key] = buildNode(key, "base");
...
}
...
 for (const [child, parents] of Object.entries(base.parent_map)) {
    for (const parent of parents) {
      const childNode = nodes[child];
      const parentNode = nodes[parent];
      const id = `${parent}_${child}`;
      edges[id] = {
        id,
        from: "base",
        parent: parentNode,
        child: childNode,
      };
      const edge = edges[id];

      childNode.parents[parent] = edge;
      parentNode.children[child] = edge;
    }
  }
```

**Comment by popcornylu** [comment]
@kentwelcome  Do u know what condition will cause this issue?

**Comment by kentwelcome** [comment]
> @kentwelcome Do u know what condition will cause this issue?

I guess the `parent_map` is a subset of nodes. If a node has no parent, it could be not existed in the `parent_map`.

**Comment by popcornylu** [comment]
I found a way to reproduce this issue in the jaffleshop


1. create a snapshot
    ```
    $ cat ./snapshots/my_snapshot.sql
    {% snapshot my_snapshot %}
        {{
            config(
                target_schema='snapshots',
                unique_key='customer_id',
                strategy='timestamp',
                updated_at='first_order'
            )
        }}
    
        select * from {{ ref('customers')}}
    {% endsnapshot %}
    ```
1. use the snapshot
    ```
    $ cat ./models/use_snapshot.sql
    select * from {{ ref('my_snapshot')}}
    ```
1. prep the env
    ```
    dbt snapshot
    dbt run
    ```
1. Run recce
    ```
    recce server
    ```


I got the error
![image](https://github.com/DataRecce/recce/assets/860633/242a3391-cae4-44e2-83ef-c92c3b69ee88)

**Comment by popcornylu** [comment]
BTW, there is no reason that we should exclude the snapshot node. We should add it to the lineage.

**Comment by popcornylu** [comment]
Fixed in v0.22.3
